lib/src/theme/color_extensions.dart

```diff
@@ -0,0 +1,283 @@
+
+    return hslColor.withLightness(percentage).toColor();
+  }
+
+  /// Gets lightness of color.
+  double get lightness => HSLColor.fromColor(this).lightness;
+
+  /// Calculates the contrast ratio between the current color and the given color [b].
+  ///
+  /// The contrast ratio is computed based on the relative luminance of the two colors
+  /// and is useful for ensuring readability and accessibility of text on backgrounds.
+  ///
+  /// Returns the computed contrast ratio.
+  double contrastRatio(Color b) {
+    final luminanceA = computeLuminance();
+    final luminanceB = b.computeLuminance();
+    if (luminanceA > luminanceB) {
+      return (luminanceA + 0.05) / (luminanceB + 0.05);
+    }
+    return (luminanceB + 0.05) / (luminanceA + 0.05);
+  }
+
+  /// Adjusts the color to a specific contrast target.
+  ///
+  /// Parameters:
+  /// - [targetContrast] The target contrast for the color adjustment.
+  /// - [on] The color to compare contrast against. Defaults to [Colors.white].
+  /// - [maxIterations] The maximum number of iterations for color adjustment. Defaults to 1000 iterations.
+  /// - [epsilon] The difference tolerance for the contrast ratio. Defaults to 0.1.
+  ///
+  /// Returns:
+  /// - A new color with adjusted contrast close to the target contrast.
+  Color adjustContrast(
+    double targetContrast, {
+    required Color on,
+    int maxIterations = 1000,
+    double epsilon = 0.1,
+  }) {
+    int iteration = 0;
+    Color adjustedColor = this;
+    double adjustmentValue;
+
+    while (iteration < maxIterations) {
+      final currentContrast = adjustedColor.contrastRatio(on);
+
+      if (currentContrast == targetContrast || (currentContrast - targetContrast).abs() < epsilon) {
+        break;
+      }
+
+      final hslColor = HSLColor.fromColor(adjustedColor);
+      adjustmentValue = (currentContrast - targetContrast).abs() * 0.02;
+
+      var newLightness = (currentContrast < targetContrast)
+          ? hslColor.lightness - adjustmentValue
+          : hslColor.lightness + adjustmentValue;
+
+      newLightness = newLightness.clamp(0.0, 1.0);
+
+      adjustedColor = hslColor.withLightness(newLightness).toColor();
+      iteration++;
+    }
+
+    return adjustedColor;
+  }
+
+  ///Generates a swatch with adjusted color contrasts.
+  ///
+  ///Parameters:
+  ///- [primary] The primary color index. Defaults to [kZetaSwatchPrimaryIndex].
+  ///- [targetContrasts] A map of target contrast values for each shade. Defaults to [kZetaSwatchTargetContrasts].
+  ///
+  ///Returns:
+  ///- A map of color shades with adjusted color to the target contrasts.
+  Map<int, Color> generateSwatch({
+    int primary = kZetaSwatchPrimaryIndex,
+    Map<int, double> targetContrasts = kZetaSwatchTargetContrasts,
+    Color background = Colors.white,
+  }) {
+    final swatch = <int, Color>{};
+    final adjustedPrimary = adjustContrast(targetContrasts[primary]!, on: background);
```

Comment:
is `targetContrasts[primary]!` null-safe?

Reply:
I've included assertions to confirm values aren't null before we utilize them.